File: reporters/golang/tt/reporter/ws-reporter.go
package reporter

import (
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"sync"
	"time"

	"crypto/sha256"

	"github.com/gorilla/websocket"
	"github.com/pietrzakacper/tracethat.dev/reporters/golang/tt/config"
)

type Reporter interface {
	open() error
	registerEvent(payload interface{}) error
}

func (r *webSocketReporter) sendRegisterEventMessage(payload interface{}) error {
	msg, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	encryptedMsg, err := encrypt(string(msg), config.Load().Token)
	if err != nil {
		return err
	}

	r.wsMutex.Lock()
	defer r.wsMutex.Unlock()

	if r.ws == nil {
		return errors.New("socket closed")
	}

	r.ws.SetWriteDeadline(time.Now().Add(3 * time.Second))

	return r.ws.WriteMessage(websocket.TextMessage, []byte(encryptedMsg))
}

type webSocketReporter struct {
	ws      *websocket.Conn
	wsMutex *sync.Mutex

	inactivityTimer *time.Timer
}

var instance *webSocketReporter

func GetWebSocketReporterInstance() *webSocketReporter {
	if instance != nil {
		return instance
	}

	r := &webSocketReporter{
		wsMutex:         &sync.Mutex{},
		inactivityTimer: nil,
	}

	instance = r
	return r
}

const inactivityDeadline = 30 * time.Second

func (r *webSocketReporter) open() error {
	r.wsMutex.Lock()
	defer r.wsMutex.Unlock()

	if r.ws != nil {
		return nil
	}

	if config.Load().Token == "" {
		time.Sleep(100 * time.Millisecond)
	}

	token := config.Load().Token
	if token == "" {
		log.Println("[tracethat.dev] Couldn't open a socket, no token provided")
		return fmt.Errorf("no token provided")
	}

	h := sha256.New()
	h.Write([]byte(token))
	roomId := fmt.Sprintf("%x", h.Sum(nil))

	url := config.Load().ServerUrl + "/api/report?roomId=" + roomId

	ws, _, err := websocket.DefaultDialer.Dial(url, nil)

	if err != nil {
		return err
	}

	r.ws = ws

	return nil
}

func (r *webSocketReporter) RegisterEvent(payload interface{}) error {
	impl := func() error {
		err := r.open()

		if err != nil {
			return err
		}

		err = r.sendRegisterEventMessage(payload)

		if err != nil {
			log.Printf("cleanup - register event err: %v\n", err)
			r.cleanup()
			return err
		}

		if r.inactivityTimer == nil {
			r.inactivityTimer = time.AfterFunc(inactivityDeadline, func() {
				r.cleanup()
			})
		} else {
			r.inactivityTimer.Reset(inactivityDeadline)
		}

		return nil
	}

	err := impl()

	if err == nil {
		return nil
	}

	err = impl()

	return err
}

func (r *webSocketReporter) cleanup() {
	r.wsMutex.Lock()
	defer r.wsMutex.Unlock()

	if r.ws != nil {
		r.ws.Close()
		r.ws = nil
	}
}

File: reporters/javascript/tracethat.dev/src/reporter/ws-reporter.ts
import WebSocket from "isomorphic-ws";
import { onConfigChange, runtimeConfig } from "../runtime-config";
import { Reporter } from "./interface";
import { sleep, stringify } from "../utils";
import { encrypt, sha256 } from "../crypto";
import { ProcessExitBlocker } from "../process-exit-blocker";
import { log } from "../logger";

async function sendRegisterEventMessage(ws: WebSocket, payload: any) {
  const msg = await encrypt(stringify(payload), runtimeConfig.token!);
  
  return new Promise((res, rej) => {
    ws.send(msg, (err) => {
      if (err) {
        rej(err);
      } else {
        res(undefined);
      }
    });
  })
}

const INACTIVITY_TIMEOUT_MS = 30 * 1000;
const MAX_PARALLEL_CALLS = 10;

class WebSocketReporter implements Reporter {
  private ws: Promise<WebSocket> | null = null;
  private cleanupTimeoutRef: NodeJS.Timeout | null = null;
  private parellelCalls = 0;

  constructor() {
    ProcessExitBlocker.instance.onShouldExit(() => {
      this.cleanup(true);
    });

    onConfigChange(() => {
      log("Config changed")
      this.cleanup();
    });
  }

  private async open(): Promise<WebSocket> {
    const callId = `[${Math.random().toString(36).substring(7)}] `;
    log(callId + "open() called ")

    if (this.ws) {
      log(callId + "returned existing ws")
      return this.ws;
    }

    if (!runtimeConfig.token) {
      log(callId + "sleeping for token")
      await sleep(100);
    }

    if (!runtimeConfig.token) {
      log(callId + "did not find token, rejecting")

      console.error("[tracethat.dev] Couldn't open a socket, no token provided");
      return Promise.reject("No token provided");
    }

    log(callId + "creating new connection with token: " + runtimeConfig.token)

    return (this.ws = sha256(runtimeConfig.token).then(roomId => new Promise((resolve, reject) => {
      let url = `${runtimeConfig.serverUrl}/api/report?roomId=${roomId}`
      if(runtimeConfig.debug) {
        url += `&debugToken=${runtimeConfig.token}`
      }

      const ws = new WebSocket(url);
      ws.onopen = function () {
        // The WebSocket type doesn't expose this property, but every Socket has it
        // we need to unref it so that it doesn't stop the Node.JS process from exiting
        this._socket?.unref();
        resolve(ws);
      };

      ws.onerror = (err) => {
        ws.close();
        this.ws = null;
        reject(err);
      };
    })))
  }

  private async cleanup(force = false) {
    this.cleanupTimeoutRef && clearTimeout(this.cleanupTimeoutRef);
    if (this.ws) {
      return this.ws.then((ws) => {
        ws.close();
        force && ws.terminate?.();
        this.ws = null;
        this.parellelCalls = 0;
      });
    }
  }

  private scheduleCleanup() {
    this.cleanupTimeoutRef && clearTimeout(this.cleanupTimeoutRef);
    this.cleanupTimeoutRef = setTimeout(() => {
      this.cleanup();
    }, INACTIVITY_TIMEOUT_MS);
  }

  async registerEvent(payload: any) {
    const impl = async () => {
      while (this.parellelCalls >= MAX_PARALLEL_CALLS) {
        log(`[${payload.callId}] waiting for parellel calls to go down`)
        await sleep(10);
      }
      this.parellelCalls++;

      try {
        const ws = await this.open();

        if (ws.readyState !== WebSocket.OPEN) {
          throw new Error("Socked closed, failed to send an event");
        }

        await sendRegisterEventMessage(ws, payload);

        this.scheduleCleanup();
      } catch (e) {
        await this.cleanup();
        throw e;
      } finally {
        this.parellelCalls--;
      }
    };

    try {
      await impl();
      return;
    } catch (e) {
      // first one failed, try again
    }

    try {
      await impl();
    } catch (e) {
      // not much a user can do here
      console.error(e);
    }
  }
}

export const reporter = new WebSocketReporter();
